src/runtime/opengl.cpp

```diff
@@ -788,76 +810,81 @@ WEAK int halide_opengl_device_malloc(void *user_context, buffer_t *buf) {
+            global_state.GetTexLevelParameteriv(GL_TEXTURE_2D, 0, GL_TEXTURE_HEIGHT, &height);
+            if (global_state.CheckAndReportError(user_context, "halide_opengl_device_malloc binding texture (GLES3)")) {
+                return 1;
+            }
+            if (width < buf->extent[0] || height < buf->extent[1]) {
 
-            error(user_context)
-                << "Existing texture is smaller than buffer. "
-                << "Texture size: " << width << "x" << height
-                << ", buffer size: " << buf->extent[0] << "x" << buf->extent[1];
+                error(user_context)
+                    << "Existing texture is smaller than buffer. "
+                    << "Texture size: " << width << "x" << height
+                    << ", buffer size: " << buf->extent[0] << "x" << buf->extent[1];
 
-            return 1;
-        }
-#endif
-    } else {
-        if (buf->extent[3] > 1) {
-            error(user_context) << "3D textures are not supported";
-            return 1;
-        }
+                return 1;
+            }
+    #endif
+        } else {
+            if (buf->extent[3] > 1) {
+                error(user_context) << "3D textures are not supported";
+                return 1;
+            }
 
-        // Generate texture ID
-        global_state.GenTextures(1, &tex);
-        if (global_state.CheckAndReportError(user_context, "halide_opengl_device_malloc GenTextures")) {
-            return 1;
-        }
+            // Generate texture ID
+            global_state.GenTextures(1, &tex);
+            if (global_state.CheckAndReportError(user_context, "halide_opengl_device_malloc GenTextures")) {
+                return 1;
+            }
 
-        // Set parameters for this texture: no interpolation and clamp to edges.
-        global_state.BindTexture(GL_TEXTURE_2D, tex);
-        global_state.TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
-        global_state.TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
-        global_state.TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
-        global_state.TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
-        if (global_state.CheckAndReportError(user_context, "halide_opengl_device_malloc binding texture")) {
-            return 1;
-        }
+            // Set parameters for this texture: no interpolation and clamp to edges.
+            global_state.BindTexture(GL_TEXTURE_2D, tex);
+            global_state.TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+            global_state.TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+            global_state.TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+            global_state.TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+            if (global_state.CheckAndReportError(user_context, "halide_opengl_device_malloc binding texture")) {
+                return 1;
```

Comment:
This code should call DeleteTextures in the error paths to free the texture names.